flang/lib/Optimizer/CodeGen/CodeGen.cpp

```diff
@@ -3505,6 +3505,18 @@ struct ZeroOpConversion : public FIROpConversion<fir::ZeroOp> {
   }
 };
 
+class DeclareOpConversion : public FIROpConversion<fir::DeclareOp> {
```

Comment:
While experimenting with delayed privation for `hlfir` (so far I have been experimenting with `fir`), I came across a small issue which is that `fir.declare` is not support when converting from `fir` to `mlir` dialect. That's why I had to copy this conversion pattern from `PreCGRewrite.cpp`. Since `DeclareOpConversion` does not convert to op in the `fir::cg` dialect, I think it would be nice to move it here to make conversion from `fir` to `llvm` dialect more complete.

Do you folks have any input on this?

Reply:
Thanks @tblah for taking a look. I was referring to lowering using `fir-opt --fir-to-llvm-ir` for testing purposes. I expected that the `fir-opt --fir-to-llvm-ir` should provide a more-or-less complete conversion pipeline from `fir` to `llvm`. This is mostly to make testing and experimenting easier.

I am a bit of newbie to the project so I might not be using the different tools and their flags properly yet.

Reply:
No `fir-oppt --fir-to-llvm-ir` is not sufficient. The FIR pass pipeline is quite long (see https://github.com/llvm/llvm-project/blob/main/flang/test/Driver/mlir-pass-pipeline.f90). A few of these are optimizations (e.g. SimplifyIntrinsics, CSE, AliasTags), but several different passes are needed to convert FIR into LLVM (even more for HLFIR).

Roughly, to get from FIR to LLVM you need the passes in `createDefaultFIRCodeGenPassPipeline`: https://github.com/llvm/llvm-project/blob/ac3bd2bd530127786741bd9f164d66a3c3f40961/flang/include/flang/Tools/CLOptions.inc#L302

From the command line you can pipe FIR into `tco` and it will output LLVM IR. `fir-opt` is more for running individual passes. The help text is a bit hard to follow, you can find `tco` specific options at the top of this file https://github.com/llvm/llvm-project/blob/main/flang/tools/tco/tco.cpp#L41

For example, `./bin/bbc -emit-hlfir -o - file.f90 | ./bin/tco` is a bit like running `flang-new -fc1 -O2 -emit-llvm -o - file.f90`